src/Rector/Class_/RenameClassRector.php

```diff
@@ -176,4 +203,96 @@ private function isValidClassNameChange(Node $node, string $newName, Class_ $cla
         }
         return ! (in_array($node, $classNode->implements, true) && class_exists($newName));
     }
+
+    private function refactorNamespaceNode(Namespace_ $node): ?Node
+    {
+        $name = $this->getName($node);
+        if ($name === null) {
+            return null;
+        }
+
+        $classNode = $this->getClassOfNamespaceToRefactor($node);
+        if (! $classNode) {
+            return null;
+        }
+
+        $newClassFqn = $this->oldToNewClasses[$this->getName($classNode)];
+        $newNamespace = $this->getNamespacePart($newClassFqn);
+
+        // Renaming to class without namespace (example MyNamespace\DateTime -> DateTimeImmutable)
+        if (! $newNamespace) {
+            $classNode->name = $newClassFqn;
+
+            return $classNode;
+        }
+
+        $node->name = new Name($newNamespace);
+
+        return $node;
+    }
+
+    private function getClassOfNamespaceToRefactor(Namespace_ $namespace): ?ClassLike
+    {
+        $foundClass = $this->betterNodeFinder->findFirst($namespace, function (Node $node) {
+            if (! $node instanceof ClassLike) {
+                return false;
+            }
+
+            return isset($this->oldToNewClasses[$this->getName($node)]);
+        });
+
+        if ($foundClass instanceof ClassLike) {
+            return $foundClass;
+        }
+
+        return null;
+    }
+
+    private function refactorClassLikeNode(ClassLike $classLike): ?Node
+    {
+        $name = $this->getName($classLike);
+        if ($name === null) {
+            return null;
+        }
+
+        $newName = $this->oldToNewClasses[$name] ?? null;
+        if (! $newName) {
+            return null;
+        }
+
+        if (in_array($name, $this->alreadyProcessed, true)) {
```

Comment:
Here small comment might help.
```suggestion
        // prevents re-iterating same class in endless loop
        if (in_array($name, $this->alreadyProcessed, true)) {
```
It's pretty WTF if one doesn't know how tree traversion works